packages/cli/src/commands/generate.ts

```diff
@@ -91,14 +100,77 @@ export async function generate(opts: GenerateOptions) {
   }
 
   await replacePackageJson(descriptorsDir, hash)
-  await compileCodegen(descriptorsDir)
+  const cleanCodegen = await compileCodegen(descriptorsDir)
+  if (cleanCodegen) {
+    await tagGenerated(descriptorsDir, chains, whitelist)
+  }
   await fs.rm(descriptorSrcDir, { recursive: true })
   if (!config.options?.noDescriptorsPackage) {
     await runInstall()
     await flushBundlerCache()
   }
 }
 
+async function tagGenerated(
+  descriptorsDir: string,
+  chains: {
+    key: string
+    metadataRaw: Uint8Array
+  }[],
+  whitelist: string[] | null,
+) {
+  const filePath = join(descriptorsDir, "generated.json")
+
+  await fs.writeFile(
+    filePath,
+    JSON.stringify({
+      cliVersion,
+      whitelist,
+      chains: Object.fromEntries(
+        chains.map(({ key, metadataRaw }) => [
+          key,
+          toHex(Blake2128(metadataRaw)),
+        ]),
+      ),
+    }),
+  )
+}
+async function alreadyGenerated(
+  descriptorsDir: string,
+  chains: {
+    key: string
+    metadataRaw: Uint8Array
+  }[],
+  whitelist: string[] | null,
+) {
+  const generatedJsôn = join(descriptorsDir, "generated.json")
+  if (!existsSync(generatedJsôn)) return false
+
+  try {
+    const generated = JSON.parse(
+      await fs.readFile(generatedJsôn, {
+        encoding: "utf-8",
+      }),
+    )
+    if (
+      (generated.whitelist ?? ["*"]).join(",") !=
+        (whitelist ?? ["*"]).join(",") ||
+      generated.cliVersion !== cliVersion
+    )
```

Comment:
nit: I'd sort whitelists first, since order does not matter. It is a corner case, but it is "free" to solve.

```suggestion
    whitelist?.sort()
    generate.whitelist?.sort()
    if (
      (generated.whitelist ?? ["*"]).join(",") !=
        (whitelist ?? ["*"]).join(",") ||
      generated.cliVersion !== cliVersion
    )
```